File: src/tabshifter/valueobjects/Split.java
package tabshifter.valueobjects;

import java.util.ArrayList;
import java.util.List;

import static java.lang.Math.max;

public class Split extends LayoutElement {
	/**
	 * Note that IntelliJ Splitter has reverse meaning of orientation.
	 */
	public enum Orientation {
		/**
		 * Most lines and editors are "vertical".
		 * ┌─┬─┐
		 * │ │ │
		 * └─┴─┘
		 */
		vertical,
		/**
		 * Most lines and editors are "horizontal".
		 * ┌───┐
		 * ├───┤
		 * └───┘
		 */
		horizontal
	}

	public final LayoutElement first;
	public final LayoutElement second;
	public final Orientation orientation;


	public Split(LayoutElement first, LayoutElement second, Orientation orientation) {
		this.first = first;
		this.second = second;
		this.orientation = orientation;
	}

	@Override public Size size() {
		if (orientation == Orientation.vertical) {
			return new Size(
				first.size().width + second.size().width,
				max(first.size().height, second.size().height)
			);
		} else {
			return new Size(
				max(first.size().width, second.size().width),
				first.size().height + second.size().height
			);
		}
	}

	@Override public String toString() {
		return "Split(" + orientation + " " + first + ", " + second + ")";
	}

	public static List<Split> allSplitsIn(LayoutElement rootElement) {
		final List<Split> result = new ArrayList<>();
		traverse(rootElement, element -> {
			if (element instanceof Split) {
				result.add((Split) element);
			}
			return true;
		});
		return result;
	}
}

File: src/tabshifter/layout/Split.kt
package tabshifter.layout

import kotlin.math.max

open class Split(
    val first: LayoutElement,
    val second: LayoutElement,
    val orientation: Orientation
): LayoutElement() {

    /**
     * Note that IntelliJ Splitter has reverse meaning of orientation.
     */
    enum class Orientation {
        /**
         * Most lines and editors are "vertical".
         * ┌─┬─┐
         * │ │ │
         * └─┴─┘
         */
        vertical,

        /**
         * Most lines and editors are "horizontal".
         * ┌───┐
         * ├───┤
         * └───┘
         */
        horizontal
    }

    override val size =
        if (orientation == Orientation.vertical) Size(
            width = first.size.width + second.size.width,
            height = max(first.size.height, second.size.height)
        ) else Size(
            width = max(first.size.width, second.size.width),
            height = first.size.height + second.size.height
        )

    override fun toString() =
        "Split($orientation $first, $second)"
}